### PR TITLE
Fix copy of null MS

### LINKED ISSUE #1135: Copying an empty MeasurementSet gives an exception
The following lines will make casacore throw a TableError:
```
const casacore::MeasurementSet ms;
const casacore::MeasurementSet copy(ms);
```
I didn't expect an exception from this simple code. Instead, I expected that both `ms.isNull()` and `copy.isNull()` would be true now.

### PR DESCRIPTION
**Pull request by gervandiepen**
The MS copy constructor and assignment now test for a null MS object.

Close #1135

### CODE REVIEW

File: ms/MeasurementSets/MeasurementSet.cc [hunk 1]
```diff
@@ -261,27 +261,30 @@ MeasurementSet::MeasurementSet(const MeasurementSet &other)
...
       initRefs();
+    }
   }
 }
 
 MeasurementSet::~MeasurementSet()
 {
 // check to make sure that this MS is still valid
+  if (! isNull()) {
     if (!hasBeenDestroyed_p  &&  !validate()) {
```
Comment (mnijhuis-tos) [inline ms/MeasurementSets/MeasurementSet.cc:288]: Optional: You could combine the two if statements into one statement.

File: ms/MeasurementSets/MeasurementSet.cc [hunk 2]
```diff
@@ -261,27 +261,30 @@ MeasurementSet::MeasurementSet(const MeasurementSet &other)
   hasBeenDestroyed_p(False)
```
Comment (mnijhuis-tos) [inline ms/MeasurementSets/MeasurementSet.cc:261]: The default constructor, which yields a Null MS, initializes `hasBeenDestroyed_p` to True. Can you initialize `hasBeenDestroyed_p` to `other.isNull()` here? Then you also don't need to modify the destructor.

File: ms/MeasurementSets/MeasurementSet.cc [hunk 3]
```diff
@@ -308,11 +312,15 @@ MeasurementSet::operator=(const MeasurementSet &other)
 	mrsDebugLevel_p = other.mrsDebugLevel_p;
 	memoryResidentSubtables_p = other.memoryResidentSubtables_p;
 
-	copySubtables (other);
+        if (! (isNull()  ||  other.isNull())) {
```
Comment (mnijhuis-tos) [inline ms/MeasurementSets/MeasurementSet.cc:315]: Why do you need the `isNull()` check? It looks like checking `if (!other.isNull())` should be enough here.

File: ms/MeasurementSets/MeasurementSet.cc [hunk 4]
```diff
@@ -308,11 +312,15 @@ MeasurementSet::operator=(const MeasurementSet &other)
...
+          copySubtables (other);
+        }
```
Comment (mnijhuis-tos) [inline ms/MeasurementSets/MeasurementSet.cc:317]: GitHub shows different indentation levels in this function, which is confusing.

**Review by mnijhuis-tos — CHANGES_REQUESTED** [review]

File: ms/MeasurementSets/MeasurementSet.cc (hunk 2 above)
Comment (mnijhuis-tos) [reply ms/MeasurementSets/MeasurementSet.cc:261]: I see you now set `hasBeenDestroyed_p` to `other.hasBeenDestroyed_p` if `other` is not null. I think you can directly initialize `hasBeenDestroyed_p` to `other.hasBeenDestroyed_p` here instead, which is a littile simpler.

**Review by mnijhuis-tos — CHANGES_REQUESTED** [review]

**Comment by gervandiepen** [comment]
I think it's better to assign hasBeenDestroyed_p later because an exception might be thrown in the constructor in which case the destructor should not destroy too much.

**Comment by mnijhuis-tos** [comment]
> I think it's better to assign hasBeenDestroyed_p later because an exception might be thrown in the constructor in which case the destructor should not destroy too much.

If an exception occurs in the constructor, the destructor is never called, since destructors are only called for completely initialized objects.

**Review by mnijhuis-tos — APPROVED** [review]

**Pull request merged**